### DISCUSSION TITLE
.NET 9 Preview 7

### DISCUSSION THREAD

**Discussion by jamesmontemagno** [dotnet/core#9448] [News]
We are pleased to announce the release of .NET 9 Preview 7. 

This release contains the following improvements:

## **📚Libraries**

- [Removal of `BinaryFormatter` is complete](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#removal-of-binaryformatter-is-complete)
- [Enumerate over `ReadOnlySpan<char>.Split()` segments](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#enumerate-over-readonlyspancharsplit-segments)
- [`Debug.Assert` now reports assert condition, by default.](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#debugassert-now-reports-assert-condition-by-default)
- [Compression APIs now use `zlib-ng`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#compression-apis-now-use-zlib-ng)
- [`Guid.CreateVersion7` enables creating GUIDs with a natural sort order](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#guidcreateversion7-enables-creating-guids-with-a-natural-sort-order)
- [`Interlocked.CompareExchange` for more types](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#interlockedcompareexchange-for-more-types)
- [AES-GCM and ChaChaPoly1305 algorithms enabled for iOS/tvOS/MacCatalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#aes-gcm-and-chachapoly1305-algorithms-enabled-for-iostvosmaccatalyst)
- [Changes to X.509 Certificate Loading](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#changes-to-x509-certificate-loading)
- [Support for XPS documents from XPS virtual printer](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#support-for-xps-documents-from-xps-virtual-printer)
- [Marking `Tensor<T>` as `Experimental`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md#marking-tensort-as-experimental)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/libraries.md)

## **⏱️Runtime**

- [ARM64 SVE Support](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/runtime.md#arm64-sve-support)
- [Post-Indexed Addressing on ARM64](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/runtime.md#post-indexed-addressing-on-arm64)
- [Strength Reduction in Loops](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/runtime.md#strength-reduction-in-loops)
- [Object Stack Allocation for Boxes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/runtime.md#object-stack-allocation-for-boxes)
- [GC Dynamic Adaptation To Application Sizes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/runtime.md#gc-dynamic-adaptation-to-application-sizes)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/runtime.md)

## **🛠️ SDK**
* [Container publishing improvements for insecure registries](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/sdk.md#container-publishing-improvements-for-insecure-registries)
* [More consistent environment variables for container publishing](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/sdk.md#more-consistent-environment-variables-for-container-publishing)
* [Introduction of Workload Sets for more control over workloads](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/sdk.md#introduction-of-workload-sets-for-more-control-over-workloads)
* [Mitigating analyzer mismatch issues aka 'torn SDK'](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/sdk.md#mitigating-analyzer-mismatch-issues-aka-torn-sdk)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/sdk.md)

## **C#**
- [Prioritize better overloads with `OverloadResolutionPriority` attribute](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/csharp.md#prioritize-better-overloads-with-overloadresolutionpriority-attribute)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/csharp.md)

## **🌐 ASP.NET Core**
- [SignalR supports trimming and Native AOT](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#signalr-supports-trimming-and-native-aot)
- [Microsoft.AspNetCore.OpenApi supports trimming and Native AOT](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#microsoftaspnetcoreopenapi-supports-trimming-and-native-aot)
- [Improvements to transformer registration APIs in Microsoft.AspNetCore.OpenApi](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#improvements-to-transformer-registration-apis-in-microsoftaspnetcoreopenapi)
- [Call `ProducesProblem` and `ProducesValidationProblem` on route groups](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#call-producesproblem-and-producesvalidationproblem-on-route-groups)
- [Construct `Problem` and `ValidationProblem` result types with `IEnumerable<KeyValuePair<string, object?>>` values](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#construct-problem-and-validationproblem-result-types-with-ienumerablekeyvaluepairstring-object-values)
- [`OpenIdConnectHandler` support for Pushed Authorization Requests (PAR)](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#openidconnecthandler-support-for-pushed-authorization-requests-par)
- [Data Protection support for deleting keys](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#data-protection-support-for-deleting-keys)
- [Customize Kestrel named pipe endpoints](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#customize-kestrel-named-pipe-endpoints)
- [Improved Kestrel connection metrics](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#improved-kestrel-connection-metrics)
- [Opt-out of HTTP metrics on certain endpoints and requests](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#opt-out-of-http-metrics-on-certain-endpoints-and-requests)
- [`ExceptionHandlerMiddleware` option to choose the status code based on the exception](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md#exceptionhandlermiddleware-option-to-choose-the-status-code-based-on-the-exception)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/aspnetcore.md)

## **📱 .NET MAUI**
* [Introduction of `HybridWebview`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#introduction-of-hybridwebview)
* [New `TitleBar` Control and `Window.TitleBar` for Windows](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-titlebar-control-and-windowtitlebar-for-windows)
* [`CollectionView` & `CarouselView` improvements with a new opt-in handler for iOS and Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#collectionview--carouselview-improvements-with-a-new-opt-in-handler-for-ios-and-mac-catalyst)
* [Ability to bring a `Window` to the foregrond with `ActivateWindow`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#activatewindow-added-to-bring-a-window-to-foreground)
* [`BackButtonBehavior` `OneWay` binding mode](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#backbuttonbehavior-oneway-binding-mode)
* [`BlazorWebView` backward compatibility host address](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#blazorwebview-backward-compatibility-host-address)
* [Native Embedding improvements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#native-embedding-improvements)
* [`MainPage` is Obsolete](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#mainpage-is-obsolete)
* [New Handler Disconnect Policy](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-handler-disconnect-policy)
* [New `ProcessTerminated` event on `WebView` Control](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-processterminated-event-on-webview-control)
* [New lifecycle methods for remote notifications on iOS & Mac Catalyst](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#new-lifecycle-methods-for-remote-notifications-on-ios--mac-catalyst)
* [Xcode Sync for CLI and Visual Studio Code](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md#xcode-sync-for-cli-and-visual-studio-code)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/dotnetmaui.md)

## Get started
To get started with .NET 9, [install the .NET 9 SDK](https://dotnet.microsoft.com/download/dotnet/9.0).

If you're on Windows using Visual Studio, we recommend installing the latest [Visual Studio 2022 preview](https://visualstudio.microsoft.com/vs/preview/). .NET 9 is now can be installed directly through the Visual Studio installer starting with Visual Studio 2022 17.12 Preview 1.

You can also use Visual Studio Code and the [C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit) extension with .NET 9.

You can find [.NET 9 Preview 7 Installer & Binaries](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/9.0.0-preview.7.md) on GitHub.

## Team Announcements & Discussions
Discuss this release with the product teams on GitHub through the announcements for this release. 

* [.NET MAUI](https://github.com/dotnet/maui/discussions/24219) 
* [ASP.NET Core](https://github.com/dotnet/aspnetcore/discussions/57312)
* [Libraries & Runtime](https://github.com/dotnet/runtime/discussions/106350) 
* [Source-build](https://github.com/dotnet/source-build/discussions/4551)

## Stay up-to-date with .NET 9
You can stay up-to-date with all the features of .NET 9 with:

* [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview)
* [What's new in ASP.NET Core](https://learn.microsoft.com/aspnet/core/release-notes/aspnetcore-9.0)
* [What's new in .NET MAUI](https://learn.microsoft.com/dotnet/maui/whats-new/dotnet-9)
* [What's new in EF Core](https://learn.microsoft.com/ef/core/what-is-new/ef-core-9.0/whatsnew)
* [Breaking Changes in .NET 9](https://learn.microsoft.com/dotnet/core/compatibility/9.0)
* [.NET 9 Releases](https://github.com/dotnet/core/blob/main/release-notes/9.0/README.md)

For .NET 9 previews we are utilizing GitHub Discussions to make announcements. For more information [read our discussion](https://github.com/dotnet/core/discussions/9131).

Stay up to date by subscribing to the [RSS feed](https://github.com/dotnet/core/discussions/categories/news.atom) for all release announcements.

**Comment by OwnageIsMagic** [comment]
[Prioritize better overloads with OverloadResolutionPriority attribute](https://github.com/dotnet/core/discussions/9448#prioritize-better-overloads-with-overloadresolutionpriority-attribute) link broken

**Reply by jamesmontemagno** [reply]
Thanks, fixed https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview7/csharp.md#prioritize-better-overloads-with-overloadresolutionpriority-attribute

**Comment by wstaelens** [comment]
THANK YOU @edwardneal ! 😄 🥳  #xps

**Comment by Firuz2006** [comment]
Go go go!

**Comment by fMichaleczek** [comment]
`dotnet workload install maui` fails with :

``
Workload installation failed: One or more errors occurred. (Version 17.5.9232-net9-p7 of package microsoft.ios.sdk.net9.0_17.5.msi.x64 is not found in NuGet feeds 
``

@rolfbjarne Sorry to bother you, but it looks like Preview 3 and 4.
https://github.com/dotnet/core/discussions/9318#discussioncomment-9510301

**Reply by jamesmontemagno** [reply]
@davidortinau as well

**Reply by dalexsoto** [reply]
Do you happen to have this feed in your nuget sources?

```xml
<add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
```

If so please remove it and try again, you can confirm this with `dotnet nuget list source`

**Reply by rolfbjarne** [reply]
@fMichaleczek can you do:

```
dotnet --info
``

and

```
dotnet worklod install maui -v:diagnostic
```

and attach the output here?

**Reply by fMichaleczek** [reply]
@rolfbjarne 
I installed the Preview 7 SDK, updated VS to the latest preview, then it was impossible to go out the error loop. (with repair, etc..)
After removing the Preview 7 SDK and Keep the VS Preview 7 SDK, this was not resolving my problem, so I remove also C:\Program Files\dotnet\sdk-manifests\9.0.100-preview.7 and force a repair.
It's now working.
It seems a conflict problem between VS SDK and SDK.

**Comment by nathan130200** [comment]
I'm mad to use C# extension types asap 👁👄👁

**Comment by tivik7** [comment]
I am getting an exception when using the preview 7 SDK in a WPF project using the new fluent.xaml. The pasted lines (line number 17-19 in the code) throws an exception:

            <ResourceDictionary.MergedDictionaries>
                <ResourceDictionary Source="pack://application:,,,/PresentationFramework.Fluent;component/Resources/Fluent.xaml" />
            </ResourceDictionary.MergedDictionaries>

Set property 'System.Windows.ResourceDictionary.Source' threw an exception.' Line number '82' and line position '18'.

with the inner exception:

IOException: Cannot locate resource 'resources/fluent.xaml'.

Workloads, Visual Studio Preview, NuGets and dotnet preview 7 all installed without errors

**Reply by jamesmontemagno** [reply]
@singhashish-wpf ^

**Reply by jeffhandley** [reply]
I created an issue for this, @singhashish-wpf: https://github.com/dotnet/wpf/issues/9580

**Reply by pchaurasia14** [reply]
@tivik7 @jeffhandley - Thank you for pointing this out. I've fixed the [documentation](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview4/wpf.md) as well.

**Comment by IanMarteens** [comment]
Neither semi-automatic properties, nor extensions, it seems. Am I wrong?

**Reply by KennethHoff** [reply]
Nor anything in regards to collection expressions or primary constructors, both of which were left partially unfinished in 8 🫠

params collections, while certainly an improvement, is effectively a non-feature now that we have collection expressions, and it's probably the biggest feature coming in C# 13. There are a few `ref struct` improvements, but those are barely used in application code.

**Reply by jamesmontemagno** [reply]
There was a roll up blog last month with an update in and around C# 13: https://devblogs.microsoft.com/dotnet/csharp-13-explore-preview-features/ 

These release notes are just specific to this release.

**Reply by KennethHoff** [reply]
Considering neither `params collection` nor any `ref struct` improvements were referenced in these release notes - and those were the things I mentioned by name - it's fairly obvious that we've already read those

**Reply by jamesmontemagno** [reply]
If you are looking for status and discussion of features check https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md

**Reply by KennethHoff** [reply]
This is out of date though, isn't it? `ref struct interfaces` are currently planned for C# 13.

There was a brief period where they were considered delayed/dropped, but not anymore

**Comment by Firuz2006** [comment]
Will dotnet 9 be able to ```dotnet new winui```?

**Comment by sirinath** [comment]
Looking for the performance improvement blog post.

**Reply by richlander** [reply]
You won't have to wait long.

**Comment by david-nyaoso** [comment]
After upgrade to dotnet  9 preview 7, can not deploy Maui App targeting dotnet 9 and building for Windows. This means that code which depended on Application Identity, for example 'Windows.ApplicationModel.Package.Current' throws an exception.  Is this by design?

**Reply by mattleibow** [reply]
Is this an existing app or is it a new one from the template?

**Reply by david-nyaoso** [reply]
Both new from template and existing app behave the same.

**Reply by mattleibow** [reply]
The new app template was switched to unpackaged by default for preview 7, so that is expected. However, for an existing app, this should be unchanged. The "switch" was just a msbuild property in the csproj in the template, so it should have no way of affecting your existing app.

Please open a new issue in dotnet/maui with a sample project that used to work in net8.0 but now does not in net9.

**Reply by david-nyaoso** [reply]
https://github.com/david-nyaoso/maui-win-package-bug

**Comment by BeepBeepBopBop** [comment]
Hello, after updating a .NET MAUI app from .net8 to the newest version (.net9 preview 7), I get the following error message when attempting to compile the Windows app:

> Improper project configuration: no AppxManifest is specified, but WindowsPackageType is not set to $(WindowsPackageType).


I have not found any solution out there and I have absolutely no idea how to move forward with this, any help would be greatly appreciated! Thanks.

**Comment by nathan130200** [comment]
Add compiler preprocessor define to detect if current working platform is big or little endian. `#if PLATFORM_LITTLE_ENDIAN` and `#if PLATFORM_BIG_ENDIAN` so we can compile code directly for big/little endian instead of using `BitConverter.IsLittleEndian` at runtime.

**Reply by KennethHoff** [reply]
I don't think this would work; The platform you compile on, and the platform you run the code on isn't necessarily the same thing.

**Reply by martincostello** [reply]
I think the JIT will optimise the branches away anyway?

**Reply by amanasifkhalid** [reply]
Correct, the JIT should optimize away conditionals that can be evaluated at compile-time. Consider the following example:
```csharp
public static void Main()
{
    if (BitConverter.IsLittleEndian)
    {
        Console.WriteLine("IsLittleEndian");
    }
    else
    {
        Console.WriteLine("IsBigEndian");
    }
}
```

When optimizing, the JIT produces the following linear code:
```
G_M000_IG01:                ;; offset=0x0000
       sub      rsp, 40

G_M000_IG02:                ;; offset=0x0004
       mov      rcx, 0x28A800068B0
       call     [System.Console:WriteLine(System.String)]
       nop

G_M000_IG03:                ;; offset=0x0015
       add      rsp, 40
       ret
```

**Reply by baronfel** [reply]
Also in general we've gone away from compiler defines where possible because they cause rebuilds - in general .NET compilations can be done in platform independent ways, and then tools/tech like the trimmer and linker specialize that platform-independent binary for the target platform's characteristics.

**Reply by nathan130200** [reply]
Oh that's right, I hadn't remembered that the compiler would optimize these calls at compile-time.

**Comment by HugoVG** [comment]
Is there any plans for Microsoft.AspNetCore.OpenAPI to support XML doc comments like the previous swagger implementation did? The attributes don't feel that intuitive to use, and some things like adding comments to why/when it returns the response type/code cannot be mapped yet. Is there any plans to add that in the future?

I rather have 
```cs
/// <summary>
/// An experimental endpoint to showcase some OpenAPI problem
/// </summary>
/// <remarks>This also shows in Swagger UI</remarks>
/// <param name="experiment">Experimental message</param>
/// <returns>A Hello world message</returns>
/// <response code="200">Everything is okay :) </response>
/// <response code="400">Something broke</response>
[HttpPost("Experiment")]
[ProducesResponseType(StatusCodes.Status200OK)]
[ProducesResponseType(StatusCodes.Status400BadRequest)]
public ActionResult<string> Test(string experiment)
``` 

over 
```cs
[HttpPost("Experiment")]
[EndpointSummary("This sadly doesn't feel intuitive")]
[EndpointDescription("This sadly doesn't feel intuitive")]
[Produces(typeof(string))]
[ProducesErrorResponseType(typeof(ProblemDetails))]
[ProducesResponseType(StatusCodes.Status200OK)]
[ProducesResponseType(StatusCodes.Status400BadRequest)]
public ActionResult<string> Test([Description("Experimental message")]string experiment)
```

**Reply by martincostello** [reply]
See https://github.com/dotnet/aspnetcore/issues/39927#issuecomment-2233634912.
